drivers/xbee/xbee.c

```diff
@@ -0,0 +1,700 @@
+    if (resp.status == 0) {
+        return 2;
+    }
+    return -EINVAL;
+}
+
+int _get_proto(xbee_t *dev, uint8_t *val, size_t max)
+{
+    if (max < sizeof(ng_nettype_t)) {
+        return -EOVERFLOW;
+    }
+    memcpy(val, &(dev->proto), sizeof(ng_nettype_t));
+    return sizeof(ng_nettype_t);
+}
+
+int _set_proto(xbee_t *dev, uint8_t *val, size_t len)
+{
+    if (len != sizeof(ng_nettype_t)) {
+        return -EINVAL;
+    }
+    memcpy(&(dev->proto), val, sizeof(ng_nettype_t));
+    return sizeof(ng_nettype_t);
+}
+
+/*
+ * Driver's "public" functions
+ */
+int xbee_init(xbee_t *dev, uart_t uart, uint32_t baudrate,
+              gpio_t reset_pin, gpio_t sleep_pin)
+{
+    uint8_t tmp[2];
+
+    /* check device and bus parameters */
+    if (dev == NULL) {
+        return -ENODEV;
+    }
+    if (uart >= UART_NUMOF) {
+        return -ENXIO;
+    }
+    /* set device driver */
+    dev->driver = &xbee_driver;
+    /* set peripherals to use */
+    dev->uart = uart;
+    dev->reset_pin = reset_pin;
+    dev->sleep_pin = sleep_pin;
+    /* set default options */
+    dev->proto = XBEE_DEFAULT_PROTOCOL;
+    dev->options = 0;
+    dev->addr_short[0] = (uint8_t)(XBEE_DEFAULT_SHORT_ADDR >> 8);
+    dev->addr_short[1] = (uint8_t)(XBEE_DEFAULT_SHORT_ADDR);
+    /* initialize buffers and locks*/
+    mutex_init(&(dev->tx_lock));
+    mutex_init(&(dev->resp_lock));
+    dev->resp_limit = 1;    /* needs to be greater then 0 initially */
+    dev->rx_count = 0;
+    /* initialize UART and GPIO pins */
+    if (uart_init(uart, baudrate, _rx_cb, _tx_cb, dev) < 0) {
+        DEBUG("xbee: Error initializing UART\n");
+        return -ENXIO;
+    }
+    if (reset_pin < GPIO_NUMOF) {
+        if (gpio_init_out(reset_pin, GPIO_NOPULL) < 0) {
+            DEBUG("xbee: Error initializing RESET pin\n");
+            return -ENXIO;
+        }
+        gpio_set(reset_pin);
+    }
+    if (sleep_pin < GPIO_NUMOF) {
+        if (gpio_init_out(sleep_pin, GPIO_NOPULL) < 0) {
+            DEBUG("xbee: Error initializing SLEEP pin\n");
+            return -ENXIO;
+        }
+        gpio_clear(sleep_pin);
+    }
+    /* if reset pin is connected, do a hardware reset */
+    if (reset_pin < GPIO_NUMOF) {
+        gpio_clear(reset_pin);
+        hwtimer_wait(HWTIMER_TICKS(RESET_DELAY));
+        gpio_set(reset_pin);
+    }
+     /* put the XBee device into command mode */
+    hwtimer_wait(HWTIMER_TICKS(ENTER_CMD_MODE_DELAY));
+    _at_cmd(dev, "+++");
+    hwtimer_wait(HWTIMER_TICKS(ENTER_CMD_MODE_DELAY));
+    /* disable non IEEE802.15.4 extensions */
+    _at_cmd(dev, "ATMM2\r");
+    /* put XBee module in "API mode without escaped characters" */
+    _at_cmd(dev, "ATAP1\r");
+    /* apply AT commands */
+    _at_cmd(dev, "ATAC\r");
+    /* exit command mode */
+    _at_cmd(dev, "ATCN\r");
+
+    /* set default short address */
+    _set_addr(dev, dev->addr_short, 2);
+    /* load long address (we can not set it, as it is read only) */
```

Comment:
I would also point out, that this is only true for this particular device, keeping in mind that this is supposed to be a reference implementation